### ISSUE TITLE
Multi-epoch training differs from single-epoch training in the first epoch when using grain with parquet

### ISSUE THREAD

**Issue by emergenz** [AI-Hypercomputer/maxtext#2543]
### Bug report

Essentially the culprit is the shuffling logic in the grain parquet codepath of maxtext:
https://github.com/AI-Hypercomputer/maxtext/blob/58dba0a78f8ca34e20a778892b2ce4409f2067d8/src/MaxText/input_pipeline/_grain_data_processing.py#L70-L84

I just quickly wanted to open the issue before I forget. If you need more information, let me know.

```bash
python3 -m MaxText.train \
  MaxText/configs/base.yml \
  run_name=${job_name}_${slurm_job_id} \
  base_output_directory=$BASE_OUTPUT_DIRECTORY \
  dataset_type=grain \
  grain_file_type=parquet \
  grain_train_files=$GRAIN_TRAIN_FILES \
  grain_eval_files=$GRAIN_EVAL_FILES \
  eval_interval=500 \
  steps=50000 \
  skip_jax_distributed_system=True
```

```bash
python3 -m MaxText.train \
  MaxText/configs/base.yml \
  run_name=${job_name}_${slurm_job_id} \
  base_output_directory=$BASE_OUTPUT_DIRECTORY \
  dataset_type=grain \
  grain_file_type=parquet \
  grain_train_files=$GRAIN_TRAIN_FILES \
  grain_eval_files=$GRAIN_EVAL_FILES \
  eval_interval=500 \
  steps=50000 \
  num_epoch=10 \
  skip_jax_distributed_system=True
```

### Logs/Output

<img width="1782" height="694" alt="Image" src="https://github.com/user-attachments/assets/83ed7a7b-0d72-487a-ac4e-567436db3e8d" />

Pink is the multi-epoch run.

### Environment Information

HEAD: `3789aa17d7d3cf88c3a4db23c97127b1c2005bfc`
Single H100-SXM5
```
Name: jax
Version: 0.7.0
Summary: Differentiate, compile, and transform Numpy code.
Home-page: https://github.com/jax-ml/jax
Author: JAX team
Author-email: jax-dev@google.com
License: Apache-2.0
Location: /fast/home/franz.srambical/crowd-pilot/maxtext/.venv/lib/python3.12/site-packages
Requires: jaxlib, ml_dtypes, numpy, opt_einsum, scipy
Required-by: aqtp, chex, clu, flax, google-jetstream, optax, orbax-checkpoint, pathwaysutils, qwix, seqio, transformer_engine_jax
---
Name: flax
Version: 0.11.2
Summary: Flax: A neural network library for JAX designed for flexibility
Home-page: https://github.com/google/flax
Author: 
Author-email: Flax team <flax-dev@google.com>
License: 
Location: /fast/home/franz.srambical/crowd-pilot/maxtext/.venv/lib/python3.12/site-packages
Requires: jax, msgpack, numpy, optax, orbax-checkpoint, PyYAML, rich, tensorstore, treescope, typing_extensions
Required-by: aqtp, clu, google-jetstream, qwix, transformer_engine_jax
---
Name: maxtext
Version: 0.1.1
Summary: MaxText is a simple, performant and scalable Jax LLM!
Home-page: 
Author: 
Author-email: 
License-Expression: Apache-2.0
Location: /fast/home/franz.srambical/crowd-pilot/maxtext/.venv/lib/python3.12/site-packages
Editable project location: /fast/home/franz.srambical/crowd-pilot/maxtext
Requires: 
Required-by: 
---
Name: grain
Version: 0.2.12
Summary: Grain: A library for loading and transforming data for ML training.
Home-page: https://github.com/google/grain
Author: 
Author-email: Grain team <grain-dev@google.com>
```

**Comment by emergenz** [comment]
Just a quick follow-up: I have confirmed that switching to ArrayRecords yields identical training curves, hence the issue is solely in the shuffling logic of the parquet codepath.

**Comment by aireenmei** [comment]
Hi @emergenz, could you clarify the issue more. Are you saying that when comparing a single epoch run to the first epoch of a 10 epoch run, in the single epoch run the data is insufficiently shuffled, so it cause local overfitting that's not seen in the first epoch of a 10 epoch run?

**Comment by emergenz** [comment]
Yes, that's my best guess as to why the loss curves differ.

**Comment by emergenz** [comment]
But it's the other way around. In the multi-epoch run (pink), there seems to be some shuffling issue.

**Comment by emergenz** [comment]
(At least that's my interpretation of the periodic loss spikes.)

**Comment by aireenmei** [comment]
@iindyk  Could you comment?

**Comment by iindyk** [comment]
oh, I think there indeed might be a bug: cycle_length in the interleave should not be set to the number of datasets -- it should be something smaller than the total number of files. For perf I think even 16 will be enough. The loss curve will be different at the edge of the epochs, but should be similar for the body of the epoch.

Could you please test if this resolves the issue?

What's happening now is that all files are repeated and interleaved together, so the same files from different epochs are interleaved in the same cycle.

**Comment by jedcheng** [comment]
@emergenz May I know how your array record file is configured? I keep getting a shape mismatch error at the last step of each epoch (the length of my dataset is not divisible by a power of 2), and I cannot find any example of training beyond 10 steps.

**Comment by aireenmei** [comment]
@jedcheng is it the "uneven completion" problem mentioned [here](https://maxtext.readthedocs.io/en/latest/guides/data_input_pipeline.html#random-access-dataset-recommended)? But this should only apply to the last few steps of the last epoch. Are you seeing errors with middle epochs as well? You can open another issue with the error message as in this thread we discuss parquet files

**Comment by jedcheng** [comment]
@aireenmei When using parquet files, the mismatch happened at ~90% of the first epoch, both single and multi-host. Thus, I initially suspected something to do with this issue. But after switching to array records, the mismatch happens at exactly the last step. I think it is more about the sampling method when loading sequential parquet files.

The mismatch happens at the last step of the second epoch when enabling generate_padding_batch_train. I have filed another issue:
https://github.com/AI-Hypercomputer/maxtext/issues/2787

Thanks
